### PR TITLE
Add disallowUnknownField Option

### PR DESCRIPTION
**Pull request by kitagry**
disallowUnknownField option likes [encoding/json#Decoder.DisallowUnknownField](https://golang.org/pkg/encoding/json/#Decoder.DisallowUnknownFields)
I think this is useful for typo.
If you don't like this option, please close PR. Thanks.

### CODE REVIEW

File: decode.go [hunk 1]
```diff
@@ -481,10 +483,22 @@ func (d *Decoder) decodeStruct(dst reflect.Value, src ast.Node) error {
 	if err != nil {
 		return errors.Wrapf(err, "failed to create struct field map")
 	}
-	keyToNodeMap, err := d.keyToNodeMap(src)
+	keyToNodeMap, err := d.keyToNodeMap(src, func(iter *ast.MapNodeIter) ast.Node {
```
Comment (goccy) [inline decode.go]: How about adding a new method like the following and using it ?

```go
func (d *Decoder) keyNameToNodeMap(node ast.Node, keyOrValue func(*ast.MapNodeIter) ast.Node) (map[string]ast.Node, error) {
 ...
}

func (d *Decoder) keyNameToKeyNodeMap(src ast.Node) (map[string]ast.Node, error) {
    m, err := d.keyNameToNodeMap(src, func(iter *ast.MapNodeIter) ast.Node { return iter.Key() })
     ....
    return m, nil
}

func (d *Decoder) keyNameToValueNodeMap(src ast.Node) (map[string]ast.Node, error) {
    m, err := d.keyNameToNodeMap(src, func(iter *ast.MapNodeIter) ast.Node { return iter.Value() })
    return m, nil
}
```

**Comment by kitagry** [comment]
Thank you for the review. fixed.

File: decode.go [hunk 2]
```diff
@@ -440,6 +440,14 @@ func (d *Decoder) keyToNodeMap(node ast.Node, filter func(*ast.MapNodeIter) ast.
 	return keyToNodeMap, nil
 }
 
+func (d *Decoder) keyToKeyNodeMap(node ast.Node) (map[string]ast.Node, error) {
+	return d.keyToNodeMap(node, func(nodeMap *ast.MapNodeIter) ast.Node { return nodeMap.Key() })
```
Comment (goccy) [inline decode.go]: Would you please wrap error ?

```go
m, err := d.keyToNodeMap()
if err != nil {
    return nil, errors.Wrap(err, "...")
}
return m, nil
```

Comment (kitagry) [reply decode.go]: Oh, sorry.
Fix it soon.

**Comment by kitagry** [comment]
Fixed.

File: decode_test.go [hunk 3]
```diff
@@ -1233,3 +1233,24 @@ bar: c
...
+
+func Example_DisallowUnknownField() {
+	var v struct {
+		A string `yaml:"simple"`
+		C string `yaml:"complicate"`
+	}
+
+	const src = `---
+simple: string
+complecate: string
```
Comment (lestrrat) [inline decode_test.go]: s/complecate/complicated/

Comment (lestrrat) [reply decode_test.go]: ah, I get it. in that case 1240 should be "complicated"

File: decode.go [hunk 4]
```diff
@@ -610,6 +636,11 @@ func (d *Decoder) decodeStruct(dst reflect.Value, src ast.Node) error {
 			}
 		}
 	}
+	if len(uncalledKeys) != 0 && d.disallowUnknownField {
```
Comment (lestrrat) [inline decode.go]: s/uncalledKeys/unknownFields/ ?

**Comment by kitagry** [comment]
Thank you for the review! fixed.

File: decode.go [hunk 5]
```diff
@@ -407,7 +409,7 @@ func (d *Decoder) castToAssignableValue(value reflect.Value, target reflect.Type
 	return value
 }
 
-func (d *Decoder) keyToNodeMap(node ast.Node) (map[string]ast.Node, error) {
+func (d *Decoder) keyToNodeMap(node ast.Node, filter func(*ast.MapNodeIter) ast.Node) (map[string]ast.Node, error) {
```
Comment (goccy) [inline decode.go]: Would you please fix variable name from `filter` to `getKeyOrValueNode`

File: decode.go [hunk 6]
```diff
@@ -610,6 +636,11 @@ func (d *Decoder) decodeStruct(dst reflect.Value, src ast.Node) error {
 			}
 		}
 	}
+	if len(unknownFields) != 0 && d.disallowUnknownField {
+		for key, node := range unknownFields {
+			return errors.ErrSyntax(fmt.Sprintf("unknown field \"%s\"", key), node.GetToken())
```
Comment (goccy) [inline decode.go]: There is no need to escape with double quotes if implemented in the following way :)

```go
fmt.Sprintf(`unknown field "%s"`,  key)
```

**Comment by goccy** [comment]
LGTM !!

**Comment by kitagry** [comment]
Thank you for the review! I learned a lot.

**Pull request merged**